third_party/move/move-model/src/model.rs

```diff
@@ -2112,6 +2098,15 @@ impl<'env> ModuleEnv<'env> {
         &self.data.use_decls
     }
 
+    /// Does this module have a friend declaration for `mod_name`?
+    pub fn has_friend_named(&self, mod_name: &ModuleName) -> bool {
+        // Note: we could optimize this lookup using a hash set if needed.
```

Comment:
The old field `friend_modules: BTreeSet<ModuleId>` seems more suitable for doing this lookup.  Why did you change to a vector of new `FriendDecl` structs?

Reply:
This was actually how I started my original implementation (i.e., using `friend_modules`).

Here is why I changed it: that field was never populated. The field is part of `ModuleData`, which is constructed by visiting the modules in a topologically sorted order (with "used" modules visited first) and mapped to a `ModuleId`. Because friend decls forms the reverse of module usage (you declare a module as a friend if you expect that module to use you), the `ModuleId` for friend modules is not available when constructing `ModuleData`. Thus,  you would need to store the `FriendDecls` anyway, and do a pass over all `ModuleData` after their construction to populate the `friend_modules` (possibly at two locations in the code, `run_move_checker` and `run_spec_checker`).  Instead, I chose to just use the friend decls directly (as is, the `friend_modules` were being cloned by the getter, so we would have wanted to implement a method similar to this).

We could still populate the `friend_modules` (as redundant, but potentially more efficient in some cases, especially if there are other use cases for friend module lookups in the future), and use that underneath this method (hence the comment), as a usecase-driven optimization in the future.

Reply:
See my comments above: (a) want to keep the old function for the case model is build from bytecode only (b) have new function for v2 but have the module name resolve to module id.